File: src/main/java/mods/cybercat/gigeresque/common/util/nest/NestBuildingHelper.java
package mods.cybercat.gigeresque.common.util.nest;

import com.mojang.datafixers.util.Pair;
import mods.cybercat.gigeresque.common.block.GigBlocks;
import mods.cybercat.gigeresque.common.block.NestResinWebBlock;
import mods.cybercat.gigeresque.common.block.NestResinWebVariant;
import mods.cybercat.gigeresque.common.entity.AlienEntity;
import mods.cybercat.gigeresque.common.tags.GigTags;
import net.minecraft.core.BlockPos;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.Block;

import java.util.List;
import java.util.Random;
import java.util.concurrent.atomic.AtomicInteger;

public record NestBuildingHelper() {
    public static void tryBuildNestAround(Level level, BlockPos pos) {
        for (var x = -1; x <= 1; x++) {
            for (var z = -1; z <= 1; z++) {
                for (var y = -1; y <= 3; y++) {
                    var blockPos = pos.offset(x, y, z);
                    var nestBlockData = getNestBlockData(level, blockPos);
                    if (nestBlockData == null)
                        continue;

                    if (level.getLightEmission(pos) < 6) {
                        var resinBlock = GigBlocks.NEST_RESIN.defaultBlockState();
                        if (nestBlockData.isFloor() && !level.getBlockState(blockPos).is(GigTags.DUNGEON_BLOCKS))
                                level.setBlockAndUpdate(blockPos, resinBlock);

                        if (nestBlockData.isCorner() && !level.getBlockState(blockPos).is(GigTags.DUNGEON_BLOCKS))
                                level.setBlockAndUpdate(blockPos, GigBlocks.NEST_RESIN_WEB_CROSS.defaultBlockState());

                        if (nestBlockData.isWall() || nestBlockData.isCeiling()) {
                            var nestResinWebState = GigBlocks.NEST_RESIN_WEB.defaultBlockState().setValue(NestResinWebBlock.UP, nestBlockData.hasUpCoverage()).setValue(NestResinWebBlock.NORTH, nestBlockData.hasNorthCoverage()).setValue(NestResinWebBlock.SOUTH, nestBlockData.hasSouthCoverage()).setValue(NestResinWebBlock.EAST, nestBlockData.hasEastCoverage()).setValue(NestResinWebBlock.WEST, nestBlockData.hasWestCoverage()).setValue(NestResinWebBlock.VARIANTS,
                                    NestResinWebVariant.values()[level.getRandom().nextInt(NestResinWebVariant.values().length)]);
                            if (!level.getBlockState(blockPos).is(GigTags.DUNGEON_BLOCKS))
                                level.setBlockAndUpdate(blockPos, nestResinWebState);
                        }
                    }
                }
            }
        }
    }

    public static boolean isResinBlock(Block block) {
        return block.defaultBlockState().is(GigTags.NEST_BLOCKS);
    }

    private static NestBlockData getNestBlockData(Level world, BlockPos blockPos) {
        var actualState = world.getBlockState(blockPos);
        var actualBlock = actualState.getBlock();

        if (isResinBlock(actualBlock))
            return null;

        var upPos = blockPos.above();
        var upState = world.getBlockState(upPos);

        var downPos = blockPos.below();
        var downState = world.getBlockState(downPos);

        var northPos = blockPos.north();
        var northState = world.getBlockState(northPos);

        var southPos = blockPos.south();
        var southState = world.getBlockState(southPos);

        var eastPos = blockPos.east();
        var eastState = world.getBlockState(eastPos);

        var westPos = blockPos.west();
        var westState = world.getBlockState(westPos);

        var blockStates = List.of(Pair.of(upPos, upState), Pair.of(downPos, downState), Pair.of(northPos, northState), Pair.of(southPos, southState), Pair.of(eastPos, eastState), Pair.of(westPos, westState));

        var fullCoverage = new AtomicInteger();
        var horizontalCoverage = new AtomicInteger();
        var cornerCoverage = new AtomicInteger();

        blockStates.forEach(it -> {
            var isOpaqueCube = it.getSecond().isSolidRender(world, it.getFirst());

            if (isOpaqueCube)
                if (!isResinBlock(it.getSecond().getBlock())) {
                    if (it.getFirst() != upPos && it.getFirst() != downPos)
                        horizontalCoverage.getAndIncrement();
                    fullCoverage.getAndIncrement();
                } else
                    cornerCoverage.getAndIncrement();
        });

        var isFloor = actualState.isAir() && upState.isAir() && !isResinBlock(downState.getBlock()) && downState.isSolidRender(world, downPos) && world.getLightEmission(blockPos) < 10;
        var isWall = !isFloor && actualState.isAir() && (upState.isAir() || isResinBlock(upState.getBlock())) && (1 <= horizontalCoverage.get() && horizontalCoverage.get() <= 2) && world.getLightEmission(blockPos) < 10;
        var isCorner = actualState.isAir() && (3 <= fullCoverage.get() && fullCoverage.get() <= 5) && world.getLightEmission(blockPos) < 10;
        var isCeiling = actualState.isAir() && (downState.isAir() || isResinBlock(downState.getBlock())) && upState.isSolidRender(world, downPos) && !isCorner && world.getLightEmission(blockPos) < 10;

        return new NestBlockData(fullCoverage.get(), isCorner, isFloor, isCeiling, isWall, /* upCoverage */ upState.isSolidRender(world, upPos), /* downCoverage */ downState.isSolidRender(world, downPos), /* northCoverage */ northState.isSolidRender(world, northPos), /* southCoverage */ southState.isSolidRender(world, southPos), /* eastCoverage */ eastState.isSolidRender(world, eastPos), /* westCoverage */ westState.isSolidRender(world, westPos));
    }
}

File: src/main/kotlin/com/bvanseg/gigeresque/common/util/nest/NestBuildingHelper.kt
package com.bvanseg.gigeresque.common.util.nest

import com.bvanseg.gigeresque.common.block.Blocks
import com.bvanseg.gigeresque.common.block.NestResinWebBlock
import com.bvanseg.gigeresque.common.block.NestResinWebVariant
import com.bvanseg.gigeresque.common.entity.AlienEntity
import net.minecraft.block.Block
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World

/**
 * @author Boston Vanseghi
 */
object NestBuildingHelper {

    fun tryBuildNestAround(alien: AlienEntity) {
        for (x in -1..1) {
            for (z in -1..1) {
                for (y in -1..3) {
                    val blockPos = alien.blockPos.add(x, y, z)
                    val nestBlockData = getNestBlockData(alien.world, blockPos) ?: continue

                    if (nestBlockData.isFloor) {
                        alien.world.setBlockState(blockPos, Blocks.NEST_RESIN.defaultState)
                    }

                    if (nestBlockData.isCorner) {
                        alien.world.setBlockState(blockPos, Blocks.NEST_RESIN_WEB_CROSS.defaultState)
                    }

                    if (nestBlockData.isWall || nestBlockData.isCeiling) {
                        val nestResinWebState = Blocks.NEST_RESIN_WEB.defaultState
                            .with(NestResinWebBlock.UP, nestBlockData.upCoverage)
                            .with(NestResinWebBlock.NORTH, nestBlockData.northCoverage)
                            .with(NestResinWebBlock.SOUTH, nestBlockData.southCoverage)
                            .with(NestResinWebBlock.EAST, nestBlockData.eastCoverage)
                            .with(NestResinWebBlock.WEST, nestBlockData.westCoverage)
                            .with(NestResinWebBlock.VARIANTS, NestResinWebVariant.values().random())
                        alien.world.setBlockState(blockPos, nestResinWebState)
                    }
                }
            }
        }
    }

    fun isResinBlock(block: Block): Boolean = block == Blocks.NEST_RESIN ||
            block == Blocks.NEST_RESIN_WEB ||
            block == Blocks.NEST_RESIN_WEB_CROSS ||
            block == Blocks.NEST_RESIN_BLOCK

    private fun getNestBlockData(world: World, blockPos: BlockPos): NestBlockData? {
        val actualState = world.getBlockState(blockPos)
        val actualBlock = actualState.block

        if (isResinBlock(actualBlock)) return null

        val upPos = blockPos.up()
        val upState = world.getBlockState(upPos)

        val downPos = blockPos.down()
        val downState = world.getBlockState(downPos)

        val northPos = blockPos.north()
        val northState = world.getBlockState(northPos)

        val southPos = blockPos.south()
        val southState = world.getBlockState(southPos)

        val eastPos = blockPos.east()
        val eastState = world.getBlockState(eastPos)

        val westPos = blockPos.west()
        val westState = world.getBlockState(westPos)

        val blockStates = listOf(
            upPos to upState,
            downPos to downState,
            northPos to northState,
            southPos to southState,
            eastPos to eastState,
            westPos to westState
        )

        var fullCoverage = 0
        var horizontalCoverage = 0
        var cornerCoverage = 0

        blockStates.forEach {
            val isOpaqueCube = it.second.isOpaqueFullCube(world, it.first)

            if (isOpaqueCube) {
                if (!isResinBlock(it.second.block)) {
                    if (it.first != upPos && it.first != downPos) {
                        horizontalCoverage++
                    }
                    fullCoverage++
                } else {
                    cornerCoverage++
                }
            }
        }

        val isFloor =
            actualState.isAir && upState.isAir && !isResinBlock(downState.block) && downState.isOpaqueFullCube(
                world,
                downPos
            )
        val isWall =
            !isFloor && actualState.isAir && (upState.isAir || isResinBlock(upState.block)) && horizontalCoverage in 1..2
        val isCorner = actualState.isAir && fullCoverage in 3..5
        val isCeiling =
            actualState.isAir && (downState.isAir || isResinBlock(downState.block)) && upState.isOpaqueFullCube(
                world,
                downPos
            ) && !isCorner

        return NestBlockData(
            coverage = fullCoverage,
            isCorner = isCorner,
            isFloor = isFloor,
            isCeiling = isCeiling,
            isWall = isWall,
            upCoverage = upState.isOpaqueFullCube(world, upPos),
            downCoverage = downState.isOpaqueFullCube(world, downPos),
            northCoverage = northState.isOpaqueFullCube(world, northPos),
            southCoverage = southState.isOpaqueFullCube(world, southPos),
            eastCoverage = eastState.isOpaqueFullCube(world, eastPos),
            westCoverage = westState.isOpaqueFullCube(world, westPos)
        )
    }
}